components/input/Input.js

```diff
@@ -0,0 +1,158 @@
+import React, { Component, createElement } from 'react';
+import PropTypes from 'prop-types';
+import cx from 'classnames';
+import { IconChevronUpSmallOutline, IconChevronDownSmallOutline, IconErrorSmallOutline } from '@teamleader/ui-icons';
+import InputMetaPropTypes from './InputMetaPropTypes';
+import Counter from '../counter';
+import theme from './theme.css';
+
+export default class Input extends Component {
+  static propTypes = {
+    bold: PropTypes.bool,
+    className: PropTypes.string,
+    counter: PropTypes.number,
+    disabled: PropTypes.bool,
+    icon: PropTypes.oneOfType([PropTypes.func, PropTypes.element]),
+    iconPlacement: PropTypes.oneOf(['left', 'right']),
+    id: PropTypes.string,
+    meta: InputMetaPropTypes,
+    onBlur: PropTypes.func,
+    onChange: PropTypes.func,
+    onFocus: PropTypes.func,
+    placeholder: PropTypes.string,
+    readOnly: PropTypes.bool,
+    size: PropTypes.oneOf(['small', 'medium', 'large']),
+    step: PropTypes.number,
+    type: PropTypes.string,
+    value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
+  };
+
+  static defaultProps = {
+    iconPlacement: 'left',
+    disabled: false,
+    placeholder: '',
+    readOnly: false,
+    size: 'medium',
+    step: 1,
+  };
+
+  constructor(props) {
+    super(props);
+
+    this.isNumberInput = Boolean(props.type === 'number');
+    this.handleChange = ::this.onChange;
+    this.handleIncreaseValue = ::this.increaseValue;
+    this.handleDecreaseValue = ::this.decreaseValue;
+
+    this.state = {
+      value: this.isNumberInput ? Number(props.value) || '' : props.value || '',
+    };
+  }
+
+  componentWillUpdate(props, state) {
+    if (state.value) {
+      this.props.onChange(state.value);
+    }
+  }
+
+  onChange(event) {
+    const value = event.target.value;
+
+    this.setState(() => ({
+      value: this.isNumberInput ? Number(value) : value,
+    }));
+  }
+
+  increaseValue() {
+    this.setState((previousState, props) => ({
+      value: previousState.value ? previousState.value + props.step : props.step,
+    }));
+  }
+
+  decreaseValue() {
+    this.setState((previousState, props) => ({
+      value: previousState.value ? previousState.value - props.step : -props.step,
+    }));
+  }
+
+  renderInput() {
+    const { bold, disabled, id, onBlur, onFocus, placeholder, readOnly, step, type } = this.props;
+    const classNames = cx(theme['input'], {
+      [theme['is-disabled']]: disabled,
+      [theme['is-read-only']]: readOnly,
+      [theme['is-bold']]: bold,
+    });
+    const props = {
+      className: classNames,
+      disabled: disabled,
+      id,
+      onBlur: onBlur,
+      onChange: this.handleChange,
+      onFocus: onFocus,
+      placeholder,
+      readOnly,
+      step,
+      type,
+      value: this.state.value,
+    };
+
+    return <input {...props} />;
```

Comment:
There's no way of adding extra attributes to the input field now (for example data-attributes). I would use destructuring so to pass just all props. Something like this (didn't test it):

```js
  renderInput() {
    const { bold, disabled, readOnly, ...others } = this.props;

    const classNames = cx(theme['input'], {
      [theme['is-disabled']]: disabled,
      [theme['is-read-only']]: readOnly,
      [theme['is-bold']]: bold,
    });

    const props = {
      className: classNames,
      disabled,
      onChange: this.handleChange,
      readOnly,
      value: this.state.value,
      ...others,
    };

    return <input {...props} />;
  }
```

This will also give you the same issues as I had with the `Box` props that should be removed first because they don't have to be passed onto the input field.

Reply:
We decided to leave it as it is now, and come back to this later.

 :ok: by @lowiebenoot on 15/02/2018 17:30:45